packages/fast-usdc/src/exos/status-manager.js

```diff
@@ -1,12 +1,152 @@
+        M.undefined(),
+      ),
+      hasPendingSettlement: M.call(M.string(), M.bigint()).returns(M.boolean()),
+      settle: M.call(M.string(), M.bigint()).returns(M.undefined()),
+      view: M.call(M.string(), M.bigint())
+        .optional(M.number())
+        .returns(
+          M.or(
+            StatusManagerEntryShape,
+            M.arrayOf(StatusManagerEntryShape),
+            M.undefined(),
+          ),
+        ),
+    }),
+    {
+      /**
+       * Queue a transaction before initiating a transfer. Receive
+       * the entry index in return.
+       *
+       * @param {CctpTxEvidence} evidence
+       * @returns {number}
+       */
+      observe(evidence) {
+        // XXX like `appendToStoredArray`, but returns index. consider
+        // upstreaming the change if it doesn't affect existing consumers
+        const key = getStatusKey(evidence);
+        const entry = { ...evidence, status: TxStatus.OBSERVED };
+        if (!txs.has(key)) {
+          // XXX append to store helper
+          txs.init(key, harden([entry]));
+          return 0;
+        }
+        const current = txs.get(key);
+        // XXX do we need defensive logic to ensure multiple entries don't have
+        // the same txHash+chainId?
+        txs.set(key, harden([...current, entry]));
+        return current.length;
+      },
+      /**
+       * Mark an `OBSERVED` transaction as `ADVANCED`.
+       *
+       * @param {string} address
+       * @param {bigint} amount
+       * @param {number} index
+       */
+      advance(address, amount, index) {
+        const key = `${address}-${amount}`;
+        const mutableEntries = [...txs.get(key)];
+        if (mutableEntries[index].status !== TxStatus.OBSERVED) {
+          throw makeError(
+            `Invalid status ${q(mutableEntries[index].status)} for ${q(key)}`,
+          );
+        }
+        mutableEntries[index] = {
+          ...mutableEntries[index],
+          status: TxStatus.ADVANCED,
+        };
+        txs.set(key, harden(mutableEntries));
+      },
+      /**
+       * Find an `ADVANCED` tx waiting to be `SETTLED`
+       *
+       * @param {string} address
+       * @param {bigint} amount
+       * @returns {boolean}
+       */
+      hasPendingSettlement(address, amount) {
+        const key = `${address}-${amount}`;
+        const entries = txs.get(key);
+        const unsettledIdx = entries.findIndex(
+          ({ status }) => status === TxStatus.ADVANCED,
+        );
+        // TODO: if OBSERVED -> SETTLED is a valid transition, so if the below
+        // is false we should probably also check for `OBSERVED`
+        return unsettledIdx > -1;
+      },
+      /**
+       * Mark an `ADVANCED` transaction as `SETTLED`.
+       *
+       * TODO - does this need to support `OBSERVED` -> `SETTLED`?
+       *
+       * @param {NobleAddress} address
+       * @param {bigint} amount
+       */
+      settle(address, amount) {
+        const key = `${address}-${amount}`;
```

Comment:
The exo guard doesn't guarantee that `address` includes no `-` characters. This sets off my confusion-attack spidey-sense.

Consider...

```suggestion
        const key = JSON.stringify([address, `${amount}`]);
```

Reply:
Please see this fixup: [5226323](https://github.com/Agoric/agoric-sdk/pull/10406/commits/522632370c23502a890a3c992870d677a2a46287)

Reply:
Another alternative: put the number first

Reply:
Or split from the right? Hm

Reply:
There are no reqs currently to decompose the key, so I think the current approach is sound. If anything, `JSON.stringify` might be overkill vs simple concatenation.

This is now factored to its own function and easy to change in the future.

Reply:
> There are no reqs currently to decompose the key
Good point. It could even be a hash function. Let's leave it as is for debug-ability and we can button up before release